File: packages/swift/Sources/LumenUI/StructuredComponents.swift
import SwiftUI

public enum LumenEmptyStateLayout: Sendable {
    case compact
    case `default`
    case page
}

public struct LumenEmptyState<Actions: View>: View {
    @Environment(\.lumenTheme) private var theme

    private let actions: Actions
    private let description: LocalizedStringKey?
    private let layout: LumenEmptyStateLayout
    private let systemName: String
    private let title: LocalizedStringKey

    public init(
        _ title: LocalizedStringKey,
        systemName: String,
        description: LocalizedStringKey? = nil,
        @ViewBuilder actions: () -> Actions
    ) {
        self.init(
            title,
            systemName: systemName,
            description: description,
            layout: .default,
            actions: actions
        )
    }

    public init(
        _ title: LocalizedStringKey,
        systemName: String,
        description: LocalizedStringKey? = nil,
        layout: LumenEmptyStateLayout = .default,
        @ViewBuilder actions: () -> Actions
    ) {
        self.title = title
        self.systemName = systemName
        self.description = description
        self.layout = layout
        self.actions = actions()
    }

    public var body: some View {
        VStack(spacing: layout == .compact ? LumenSpacing.md : LumenSpacing.lg) {
            LumenIcon(systemName: systemName, size: .lg, color: theme.colors.inkMuted)
                .padding(LumenSpacing.md)
                .background(theme.colors.surfaceMuted, in: Circle())

            VStack(spacing: LumenSpacing.sm) {
                Text(title)
                    .font(.headline)
                    .foregroundStyle(theme.colors.ink)

                if let description {
                    Text(description)
                        .font(.callout)
                        .foregroundStyle(theme.colors.inkMuted)
                        .multilineTextAlignment(.center)
                        .fixedSize(horizontal: false, vertical: true)
                }
            }

            actions
        }
        .frame(maxWidth: layout == .page ? 520 : 440)
        .padding(layout == .compact ? LumenSpacing.lg : LumenSpacing.xl)
        .frame(maxWidth: .infinity, maxHeight: layout == .compact ? nil : .infinity)
        .accessibilityElement(children: .contain)
    }
}

public extension LumenEmptyState where Actions == EmptyView {
    init(
        _ title: LocalizedStringKey,
        systemName: String,
        description: LocalizedStringKey? = nil
    ) {
        self.init(title, systemName: systemName, description: description, layout: .default) {
            EmptyView()
        }
    }

    init(
        _ title: LocalizedStringKey,
        systemName: String,
        description: LocalizedStringKey? = nil,
        layout: LumenEmptyStateLayout = .default
    ) {
        self.init(title, systemName: systemName, description: description, layout: layout) {
            EmptyView()
        }
    }
}

public enum LumenErrorStateKind: Sendable {
    case error
    case offline
}

public enum LumenErrorStateLayout: Sendable {
    case compact
    case `default`
    case page
}

/// A recoverable application or content failure.
///
/// The caller owns retry behavior and should post a platform accessibility announcement when
/// replacing existing content with this state. The state itself keeps its title and recovery
/// actions discoverable without posting repeated announcements during SwiftUI updates.
public struct LumenErrorState<Graphic: View, Actions: View>: View {
    @Environment(\.lumenTheme) private var theme

    private let actions: Actions
    private let description: LocalizedStringKey?
    private let graphic: Graphic
    private let kind: LumenErrorStateKind
    private let layout: LumenErrorStateLayout
    private let reference: String?
    private let referenceLabel: LocalizedStringKey
    private let title: LocalizedStringKey

    public init(
        _ title: LocalizedStringKey,
        description: LocalizedStringKey? = nil,
        kind: LumenErrorStateKind = .error,
        layout: LumenErrorStateLayout = .default,
        reference: String? = nil,
        referenceLabel: LocalizedStringKey = "Reference",
        @ViewBuilder graphic: () -> Graphic,
        @ViewBuilder actions: () -> Actions
    ) {
        self.title = title
        self.description = description
        self.kind = kind
        self.layout = layout
        self.reference = reference
        self.referenceLabel = referenceLabel
        self.graphic = graphic()
        self.actions = actions()
    }

    public var body: some View {
        VStack(spacing: layout == .compact ? LumenSpacing.md : LumenSpacing.lg) {
            graphic

            VStack(spacing: LumenSpacing.sm) {
                Text(title)
                    .font(.headline)
                    .foregroundStyle(theme.colors.ink)
                    .accessibilityAddTraits(.isHeader)

                if let description {
                    Text(description)
                        .font(.callout)
                        .foregroundStyle(theme.colors.inkMuted)
                        .multilineTextAlignment(.center)
                        .fixedSize(horizontal: false, vertical: true)
                }

                if let reference {
                    HStack(spacing: LumenSpacing.xs) {
                        Text(referenceLabel)
                        Text(verbatim: reference)
                    }
                    .font(.caption.monospaced())
                    .foregroundStyle(theme.colors.inkSoft)
                    #if os(iOS) || os(macOS) || os(visionOS)
                    .textSelection(.enabled)
                    #endif
                }
            }

            actions
        }
        .multilineTextAlignment(.center)
        .frame(maxWidth: 520, maxHeight: layout == .page ? .infinity : nil)
        .padding(layout == .compact ? LumenSpacing.lg : LumenSpacing.xl)
        .frame(maxWidth: .infinity, maxHeight: layout == .page ? .infinity : nil)
        .accessibilityElement(children: .contain)
    }
}

#if os(iOS) || os(macOS) || os(visionOS)
public extension LumenErrorState where Graphic == LumenIllustration {
    init(
        _ title: LocalizedStringKey,
        description: LocalizedStringKey? = nil,
        kind: LumenErrorStateKind = .error,
        layout: LumenErrorStateLayout = .default,
        reference: String? = nil,
        referenceLabel: LocalizedStringKey = "Reference",
        @ViewBuilder actions: () -> Actions
    ) {
        self.init(
            title,
            description: description,
            kind: kind,
            layout: layout,
            reference: reference,
            referenceLabel: referenceLabel,
            graphic: {
                LumenIllustration(
                    variant: kind == .error ? .error : .offline,
                    size: layout == .compact ? .sm : .md
                )
            },
            actions: actions
        )
    }
}

public extension LumenErrorState where Graphic == LumenIllustration, Actions == EmptyView {
    init(
        _ title: LocalizedStringKey,
        description: LocalizedStringKey? = nil,
        kind: LumenErrorStateKind = .error,
        layout: LumenErrorStateLayout = .default,
        reference: String? = nil,
        referenceLabel: LocalizedStringKey = "Reference"
    ) {
        self.init(
            title,
            description: description,
            kind: kind,
            layout: layout,
            reference: reference,
            referenceLabel: referenceLabel,
            actions: { EmptyView() }
        )
    }
}
#endif

public struct LumenListRow<Leading: View, Content: View, Trailing: View>: View {
    private let content: Content
    private let leading: Leading
    private let trailing: Trailing

    public init(
        @ViewBuilder leading: () -> Leading,
        @ViewBuilder content: () -> Content,
        @ViewBuilder trailing: () -> Trailing
    ) {
        self.leading = leading()
        self.content = content()
        self.trailing = trailing()
    }

    public var body: some View {
        HStack(alignment: .center, spacing: LumenSpacing.md) {
            leading
                .fixedSize()

            content
                .frame(maxWidth: .infinity, alignment: .leading)

            trailing
                .fixedSize(horizontal: true, vertical: false)
        }
        .padding(.horizontal, LumenSpacing.lg)
        .padding(.vertical, LumenSpacing.md)
        .contentShape(Rectangle())
        .accessibilityElement(children: .contain)
    }
}

public extension LumenListRow where Trailing == EmptyView {
    init(
        @ViewBuilder leading: () -> Leading,
        @ViewBuilder content: () -> Content
    ) {
        self.init(leading: leading, content: content) {
            EmptyView()
        }
    }
}

public enum LumenBannerVariant: Sendable {
    case accent
    case `default`
    case destructive
    case success
    case warning
}

public struct LumenBanner<Actions: View>: View {
    @Environment(\.lumenTheme) private var theme

    private let actions: Actions
    private let description: LocalizedStringKey?
    private let onDismiss: (() -> Void)?
    private let systemName: String?
    private let title: LocalizedStringKey
    private let variant: LumenBannerVariant

    public init(
        _ title: LocalizedStringKey,
        description: LocalizedStringKey? = nil,
        systemName: String? = nil,
        variant: LumenBannerVariant = .default,
        onDismiss: (() -> Void)? = nil,
        @ViewBuilder actions: () -> Actions
    ) {
        self.title = title
        self.description = description
        self.systemName = systemName
        self.variant = variant
        self.onDismiss = onDismiss
        self.actions = actions()
    }

    public var body: some View {
        ViewThatFits(in: .horizontal) {
            horizontalContent
            verticalContent
        }
        .padding(.horizontal, LumenSpacing.lg)
        .padding(.vertical, LumenSpacing.md)
        .background(backgroundColor)
        .overlay(alignment: .leading) {
            Rectangle()
                .fill(accentColor)
                .frame(width: 3)
        }
        .overlay {
            RoundedRectangle(cornerRadius: LumenRadius.md, style: .continuous)
                .stroke(borderColor, lineWidth: 1)
        }
        .clipShape(RoundedRectangle(cornerRadius: LumenRadius.md, style: .continuous))
        .accessibilityElement(children: .contain)
    }

    private var message: some View {
        HStack(alignment: .top, spacing: LumenSpacing.md) {
            LumenIcon(
                systemName: systemName ?? defaultSystemName,
                size: .md,
                color: accentColor
            )

            VStack(alignment: .leading, spacing: LumenSpacing.xs) {
                Text(title)
                    .font(.callout.weight(.semibold))
                    .foregroundStyle(theme.colors.ink)

                if let description {
                    Text(description)
                        .font(.caption)
                        .foregroundStyle(theme.colors.inkSoft)
                        .fixedSize(horizontal: false, vertical: true)
                }
            }
            .frame(maxWidth: .infinity, alignment: .leading)
        }
    }

    private var dismissButton: some View {
        Group {
            if let onDismiss {
                LumenIconButton(
                    systemName: "xmark",
                    label: "Dismiss",
                    intent: .quiet,
                    size: .sm,
                    action: onDismiss
                )
            }
        }
    }

    private var horizontalContent: some View {
        HStack(alignment: .center, spacing: LumenSpacing.md) {
            message
            actions
            dismissButton
        }
    }

    private var verticalContent: some View {
        VStack(alignment: .leading, spacing: LumenSpacing.md) {
            HStack(alignment: .top, spacing: LumenSpacing.sm) {
                message
                dismissButton
            }
            actions
        }
    }

    private var accentColor: Color {
        switch variant {
        case .accent: theme.colors.accent
        case .default: theme.colors.brand
        case .destructive: theme.colors.danger
        case .success: theme.colors.success
        case .warning: theme.colors.warning
        }
    }

    private var backgroundColor: Color {
        variant == .default ? theme.colors.surface : accentColor.opacity(0.08)
    }

    private var borderColor: Color {
        variant == .default ? theme.colors.line : accentColor.opacity(0.32)
    }

    private var defaultSystemName: String {
        switch variant {
        case .accent: "sparkles"
        case .default: "info.circle"
        case .destructive: "xmark.octagon"
        case .success: "checkmark.circle"
        case .warning: "exclamationmark.triangle"
        }
    }
}

public extension LumenBanner where Actions == EmptyView {
    init(
        _ title: LocalizedStringKey,
        description: LocalizedStringKey? = nil,
        systemName: String? = nil,
        variant: LumenBannerVariant = .default,
        onDismiss: (() -> Void)? = nil
    ) {
        self.init(
            title,
            description: description,
            systemName: systemName,
            variant: variant,
            onDismiss: onDismiss
        ) {
            EmptyView()
        }
    }
}

public enum LumenMetricTone: Sendable {
    case accent
    case brand
    case danger
    case neutral
    case success
    case warning
}

public enum LumenStatDensity: Sendable {
    case compact
    case `default`
}

public struct LumenStat: View {
    @Environment(\.lumenTheme) private var theme

    private let detail: LocalizedStringKey?
    private let density: LumenStatDensity
    private let label: LocalizedStringKey
    private let systemName: String?
    private let tone: LumenMetricTone
    private let value: String

    public init(
        _ label: LocalizedStringKey,
        value: String,
        detail: LocalizedStringKey? = nil,
        systemName: String? = nil,
        tone: LumenMetricTone = .brand
    ) {
        self.init(
            label,
            value: value,
            detail: detail,
            density: .default,
            systemName: systemName,
            tone: tone
        )
    }

    public init(
        _ label: LocalizedStringKey,
        value: String,
        detail: LocalizedStringKey? = nil,
        density: LumenStatDensity = .default,
        systemName: String? = nil,
        tone: LumenMetricTone = .brand
    ) {
        self.label = label
        self.value = value
        self.detail = detail
        self.density = density
        self.systemName = systemName
        self.tone = tone
    }

    public var body: some View {
        VStack(alignment: .leading, spacing: density == .compact ? LumenSpacing.xs : LumenSpacing.sm) {
            if let systemName {
                LumenIcon(systemName: systemName, size: .sm, color: accentColor)
            }

            Text(value)
                .font(density == .compact ? .headline.weight(.bold) : .title2.weight(.bold))
                .monospacedDigit()
                .foregroundStyle(theme.colors.ink)

            Text(label)
                .font(.caption.weight(.semibold))
                .foregroundStyle(theme.colors.inkSoft)

            if let detail {
                Text(detail)
                    .font(.caption)
                    .foregroundStyle(theme.colors.inkMuted)
            }
        }
        .frame(maxWidth: .infinity, alignment: .leading)
        .padding(density == .compact ? LumenSpacing.md : LumenSpacing.lg)
        .background(accentColor.opacity(0.06))
        .overlay {
            RoundedRectangle(cornerRadius: LumenRadius.md, style: .continuous)
                .stroke(accentColor.opacity(0.18), lineWidth: 1)
        }
        .clipShape(RoundedRectangle(cornerRadius: LumenRadius.md, style: .continuous))
        .accessibilityElement(children: .combine)
    }

    private var accentColor: Color {
        metricColor(tone, theme: theme)
    }
}

public struct LumenGauge: View {
    @Environment(\.lumenTheme) private var theme

    private let label: LocalizedStringKey
    private let progress: LumenProgressValue
    private let systemName: String?
    private let tone: LumenMetricTone
    private let valueLabel: String

    public init(
        _ label: LocalizedStringKey,
        value: Double,
        max: Double = 100,
        valueLabel: String,
        systemName: String? = nil,
        tone: LumenMetricTone = .brand
    ) {
        self.label = label
        progress = LumenProgressValue.resolve(value: value, max: max)
        self.valueLabel = valueLabel
        self.systemName = systemName
        self.tone = tone
    }

    public var body: some View {
        VStack(spacing: LumenSpacing.md) {
            ZStack {
                Circle()
                    .stroke(theme.colors.surfaceStrong, lineWidth: 10)

                Circle()
                    .trim(from: 0, to: progress.fraction)
                    .stroke(accentColor, style: StrokeStyle(lineWidth: 10, lineCap: .round))
                    .rotationEffect(.degrees(-90))

                VStack(spacing: LumenSpacing.xs) {
                    if let systemName {
                        LumenIcon(systemName: systemName, size: .md, color: accentColor)
                    }
                    Text(valueLabel)
                        .font(.headline.monospacedDigit())
                        .foregroundStyle(theme.colors.ink)
                }
            }
            .frame(width: 108, height: 108)

            Text(label)
                .font(.caption.weight(.semibold))
                .foregroundStyle(theme.colors.inkSoft)
        }
        .accessibilityElement(children: .ignore)
        .accessibilityLabel(Text(label))
        .accessibilityValue(Text(valueLabel))
    }

    private var accentColor: Color {
        metricColor(tone, theme: theme)
    }
}

public struct LumenSectionHeader<Actions: View>: View {
    @Environment(\.lumenTheme) private var theme

    private let actions: Actions
    private let count: String?
    private let subtitle: LocalizedStringKey?
    private let title: LocalizedStringKey

    public init(
        _ title: LocalizedStringKey,
        subtitle: LocalizedStringKey? = nil,
        count: String? = nil,
        @ViewBuilder actions: () -> Actions
    ) {
        self.title = title
        self.subtitle = subtitle
        self.count = count
        self.actions = actions()
    }

    public var body: some View {
        HStack(alignment: .top, spacing: LumenSpacing.md) {
            VStack(alignment: .leading, spacing: LumenSpacing.xs) {
                HStack(alignment: .firstTextBaseline, spacing: LumenSpacing.sm) {
                    Text(title)
                        .font(.headline)
                        .foregroundStyle(theme.colors.ink)

                    if let count {
                        LumenBadge(LocalizedStringKey(count), tone: .neutral)
                            .monospacedDigit()
                    }
                }

                if let subtitle {
                    Text(subtitle)
                        .font(.caption)
                        .foregroundStyle(theme.colors.inkMuted)
                }
            }
            .frame(maxWidth: .infinity, alignment: .leading)

            actions
        }
        .accessibilityElement(children: .contain)
    }
}

public extension LumenSectionHeader where Actions == EmptyView {
    init(
        _ title: LocalizedStringKey,
        subtitle: LocalizedStringKey? = nil,
        count: String? = nil
    ) {
        self.init(title, subtitle: subtitle, count: count) {
            EmptyView()
        }
    }
}

public struct LumenStatusBar<Trailing: View>: View {
    @Environment(\.lumenTheme) private var theme

    private let message: LocalizedStringKey
    private let systemName: String?
    private let tone: LumenMetricTone
    private let trailing: Trailing

    public init(
        _ message: LocalizedStringKey,
        tone: LumenMetricTone = .neutral,
        @ViewBuilder trailing: () -> Trailing
    ) {
        self.init(message, tone: tone, systemName: nil, trailing: trailing)
    }

    public init(
        _ message: LocalizedStringKey,
        tone: LumenMetricTone = .neutral,
        systemName: String?,
        @ViewBuilder trailing: () -> Trailing
    ) {
        self.message = message
        self.tone = tone
        self.systemName = systemName
        self.trailing = trailing()
    }

    public var body: some View {
        HStack(alignment: .top, spacing: LumenSpacing.sm) {
            LumenIcon(
                systemName: systemName ?? defaultSystemName,
                size: .sm,
                color: metricColor(tone, theme: theme)
            )

            Text(message)
                .fixedSize(horizontal: false, vertical: true)
                .frame(maxWidth: .infinity, alignment: .leading)
                .layoutPriority(1)

            trailing
                .fixedSize(horizontal: true, vertical: false)
        }
        .font(.caption)
        .foregroundStyle(theme.colors.inkSoft)
        .padding(.horizontal, LumenSpacing.lg)
        .frame(minHeight: 36)
        .background(theme.colors.surfaceMuted)
        .accessibilityElement(children: .contain)
    }

    private var defaultSystemName: String {
        switch tone {
        case .accent, .brand: "info.circle.fill"
        case .danger: "xmark.octagon.fill"
        case .neutral: "circle.fill"
        case .success: "checkmark.circle.fill"
        case .warning: "exclamationmark.triangle.fill"
        }
    }
}

public extension LumenStatusBar where Trailing == EmptyView {
    init(_ message: LocalizedStringKey, tone: LumenMetricTone = .neutral) {
        self.init(message, tone: tone) {
            EmptyView()
        }
    }

    init(
        _ message: LocalizedStringKey,
        tone: LumenMetricTone = .neutral,
        systemName: String?
    ) {
        self.init(message, tone: tone, systemName: systemName) {
            EmptyView()
        }
    }
}

private func metricColor(_ tone: LumenMetricTone, theme: LumenTheme) -> Color {
    switch tone {
    case .accent: theme.colors.accent
    case .brand: theme.colors.brand
    case .danger: theme.colors.danger
    case .neutral: theme.colors.inkMuted
    case .success: theme.colors.success
    case .warning: theme.colors.warning
    }
}

File: packages/compose/src/main/kotlin/com/santi020k/lumen/StructuredComponents.kt
package com.santi020k.lumen

import androidx.compose.foundation.BorderStroke
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.BoxWithConstraints
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxHeight
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.widthIn
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.runtime.Immutable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.drawBehind
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.heading
import androidx.compose.ui.semantics.LiveRegionMode
import androidx.compose.ui.semantics.liveRegion
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp

enum class LumenBannerVariant {
    Accent,
    Default,
    Destructive,
    Success,
    Warning
}

enum class LumenMetricTone {
    Accent,
    Brand,
    Danger,
    Neutral,
    Success,
    Warning
}

enum class LumenErrorStateAnnouncement {
    Assertive,
    Off,
    Polite
}

enum class LumenErrorStateKind {
    Error,
    Offline
}

enum class LumenErrorStateLayout {
    Compact,
    Default,
    Page
}

enum class LumenEmptyStateLayout {
    Compact,
    Default,
    Page
}

enum class LumenStatDensity {
    Compact,
    Default
}

internal fun lumenErrorStateLiveRegion(
    announcement: LumenErrorStateAnnouncement
): LiveRegionMode? = when (announcement) {
    LumenErrorStateAnnouncement.Assertive -> LiveRegionMode.Assertive
    LumenErrorStateAnnouncement.Off -> null
    LumenErrorStateAnnouncement.Polite -> LiveRegionMode.Polite
}

@Immutable
data class LumenBannerPalette(
    val accent: Color,
    val background: Color,
    val border: Color
)

fun lumenMetricColor(colors: LumenColorPalette, tone: LumenMetricTone): Color = when (tone) {
    LumenMetricTone.Accent -> colors.accent
    LumenMetricTone.Brand -> colors.brand
    LumenMetricTone.Danger -> colors.danger
    LumenMetricTone.Neutral -> colors.inkMuted
    LumenMetricTone.Success -> colors.success
    LumenMetricTone.Warning -> colors.warning
}

internal fun lumenStatusBarIconName(tone: LumenMetricTone): LumenIconName = when (tone) {
    LumenMetricTone.Accent, LumenMetricTone.Brand -> LumenIconName.Info
    LumenMetricTone.Danger -> LumenIconName.OctagonX
    LumenMetricTone.Neutral -> LumenIconName.Circle
    LumenMetricTone.Success -> LumenIconName.CircleCheck
    LumenMetricTone.Warning -> LumenIconName.TriangleAlert
}

fun lumenBannerPalette(
    colors: LumenColorPalette,
    variant: LumenBannerVariant
): LumenBannerPalette {
    val tone = when (variant) {
        LumenBannerVariant.Accent -> LumenMetricTone.Accent
        LumenBannerVariant.Default -> LumenMetricTone.Brand
        LumenBannerVariant.Destructive -> LumenMetricTone.Danger
        LumenBannerVariant.Success -> LumenMetricTone.Success
        LumenBannerVariant.Warning -> LumenMetricTone.Warning
    }
    val accent = lumenMetricColor(colors, tone)

    return LumenBannerPalette(
        accent = accent,
        background = if (variant == LumenBannerVariant.Default) colors.surface else accent.copy(alpha = 0.08f),
        border = if (variant == LumenBannerVariant.Default) colors.line else accent.copy(alpha = 0.32f)
    )
}

@Composable
fun LumenEmptyState(
    title: String,
    modifier: Modifier = Modifier,
    description: String? = null,
    layout: LumenEmptyStateLayout = LumenEmptyStateLayout.Default,
    graphic: (@Composable () -> Unit)? = null,
    actions: (@Composable () -> Unit)? = null
) {
    val colors = LocalLumenTheme.current.colors

    Column(
        modifier = modifier
            .fillMaxWidth()
            .widthIn(max = if (layout == LumenEmptyStateLayout.Page) 520.dp else 440.dp)
            .padding(if (layout == LumenEmptyStateLayout.Compact) LumenSpacing.Lg else LumenSpacing.Xl),
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.spacedBy(
            if (layout == LumenEmptyStateLayout.Compact) LumenSpacing.Md else LumenSpacing.Lg
        )
    ) {
        if (graphic != null) {
            Box(
                modifier = Modifier
                    .size(56.dp)
                    .clip(CircleShape)
                    .background(colors.surfaceMuted),
                contentAlignment = Alignment.Center
            ) {
                graphic()
            }
        }

        Column(
            horizontalAlignment = Alignment.CenterHorizontally,
            verticalArrangement = Arrangement.spacedBy(LumenSpacing.Sm)
        ) {
            Text(
                text = title,
                color = colors.ink,
                style = MaterialTheme.typography.titleMedium.copy(fontWeight = FontWeight.SemiBold),
                textAlign = TextAlign.Center
            )

            if (description != null) {
                Text(
                    text = description,
                    color = colors.inkMuted,
                    style = MaterialTheme.typography.bodyMedium,
                    textAlign = TextAlign.Center
                )
            }
        }

        actions?.invoke()
    }
}

@Composable
fun LumenErrorState(
    title: String,
    modifier: Modifier = Modifier,
    description: String? = null,
    kind: LumenErrorStateKind = LumenErrorStateKind.Error,
    layout: LumenErrorStateLayout = LumenErrorStateLayout.Default,
    announcement: LumenErrorStateAnnouncement = LumenErrorStateAnnouncement.Polite,
    reference: String? = null,
    referenceLabel: String = "Reference",
    graphic: (@Composable () -> Unit)? = null,
    actions: (@Composable () -> Unit)? = null
) {
    val colors = LocalLumenTheme.current.colors
    val compact = layout == LumenErrorStateLayout.Compact
    val liveRegionMode = lumenErrorStateLiveRegion(announcement)
    val stateModifier = modifier
        .fillMaxWidth()
        .widthIn(max = 520.dp)
        .then(if (layout == LumenErrorStateLayout.Page) Modifier.fillMaxHeight() else Modifier)
        .then(
            if (liveRegionMode == null) Modifier else Modifier.semantics {
                liveRegion = liveRegionMode
            }
        )
        .padding(if (compact) LumenSpacing.Lg else LumenSpacing.Xl)

    Column(
        modifier = stateModifier,
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.spacedBy(
            if (compact) LumenSpacing.Md else LumenSpacing.Lg,
            Alignment.CenterVertically
        )
    ) {
        if (graphic != null) {
            graphic()
        } else {
            LumenIllustration(
                variant = if (kind == LumenErrorStateKind.Error) {
                    LumenIllustrationVariant.Error
                } else {
                    LumenIllustrationVariant.Offline
                },
                size = if (compact) LumenIllustrationSize.Sm else LumenIllustrationSize.Md
            )
        }

        Column(
            horizontalAlignment = Alignment.CenterHorizontally,
            verticalArrangement = Arrangement.spacedBy(LumenSpacing.Sm)
        ) {
            Text(
                text = title,
                modifier = Modifier.semantics { heading() },
                color = colors.ink,
                style = MaterialTheme.typography.titleMedium.copy(fontWeight = FontWeight.SemiBold),
                textAlign = TextAlign.Center
            )

            if (description != null) {
                Text(
                    text = description,
                    color = colors.inkMuted,
                    style = MaterialTheme.typography.bodyMedium,
                    textAlign = TextAlign.Center
                )
            }

            if (reference != null) {
                Text(
                    text = "$referenceLabel: $reference",
                    color = colors.inkSoft,
                    style = MaterialTheme.typography.labelSmall,
                    textAlign = TextAlign.Center
                )
            }
        }

        actions?.invoke()
    }
}

@Composable
fun LumenListRow(
    modifier: Modifier = Modifier,
    leading: (@Composable () -> Unit)? = null,
    trailing: (@Composable () -> Unit)? = null,
    content: @Composable () -> Unit
) {
    Row(
        modifier = modifier
            .fillMaxWidth()
            .heightIn(min = 56.dp)
            .padding(horizontal = LumenSpacing.Lg, vertical = LumenSpacing.Md),
        horizontalArrangement = Arrangement.spacedBy(LumenSpacing.Md),
        verticalAlignment = Alignment.Top
    ) {
        leading?.invoke()

        Box(modifier = Modifier.weight(1f)) {
            content()
        }

        trailing?.invoke()
    }
}

@Composable
fun LumenBanner(
    title: String,
    modifier: Modifier = Modifier,
    description: String? = null,
    variant: LumenBannerVariant = LumenBannerVariant.Default,
    dismissLabel: String = "Dismiss",
    onDismiss: (() -> Unit)? = null,
    graphic: (@Composable () -> Unit)? = null,
    actions: (@Composable () -> Unit)? = null
) {
    val colors = LocalLumenTheme.current.colors
    val palette = lumenBannerPalette(colors, variant)
    val shape = RoundedCornerShape(LumenRadius.Md)

    Surface(
        modifier = modifier,
        color = palette.background,
        contentColor = colors.ink,
        shape = shape,
        border = BorderStroke(1.dp, palette.border)
    ) {
        Box(
            modifier = Modifier.drawBehind {
                drawRect(
                    color = palette.accent,
                    size = Size(width = 3.dp.toPx(), height = size.height)
                )
            }
        ) {
            BoxWithConstraints(
                modifier = Modifier
                    .padding(horizontal = LumenSpacing.Lg, vertical = LumenSpacing.Md)
            ) {
                val message: @Composable () -> Unit = {
                    Column(verticalArrangement = Arrangement.spacedBy(LumenSpacing.Xs)) {
                    Text(
                        text = title,
                        color = colors.ink,
                        style = MaterialTheme.typography.labelLarge.copy(fontWeight = FontWeight.SemiBold)
                    )

                    if (description != null) {
                        Text(
                            text = description,
                            color = colors.inkSoft,
                            style = MaterialTheme.typography.bodySmall
                        )
                    }
                }
                }

                if (maxWidth < 520.dp) {
                    Column(verticalArrangement = Arrangement.spacedBy(LumenSpacing.Md)) {
                        Row(horizontalArrangement = Arrangement.spacedBy(LumenSpacing.Md)) {
                            graphic?.invoke()
                            Box(modifier = Modifier.weight(1f)) { message() }
                        }
                        actions?.invoke()
                        if (onDismiss != null) {
                            TextButton(onClick = onDismiss) { Text(dismissLabel, color = palette.accent) }
                        }
                    }
                } else {
                    Row(
                        horizontalArrangement = Arrangement.spacedBy(LumenSpacing.Md),
                        verticalAlignment = Alignment.CenterVertically
                    ) {
                        graphic?.invoke()
                        Box(modifier = Modifier.weight(1f)) { message() }
                        actions?.invoke()
                        if (onDismiss != null) {
                            TextButton(onClick = onDismiss) { Text(dismissLabel, color = palette.accent) }
                        }
                    }
                }
            }
        }
    }
}

@Composable
fun LumenStat(
    label: String,
    value: String,
    modifier: Modifier = Modifier,
    detail: String? = null,
    density: LumenStatDensity = LumenStatDensity.Default,
    tone: LumenMetricTone = LumenMetricTone.Brand,
    graphic: (@Composable () -> Unit)? = null
) {
    val colors = LocalLumenTheme.current.colors
    val accent = lumenMetricColor(colors, tone)
    val accessibleText = listOfNotNull(label, value, detail).joinToString(", ")

    Column(
        modifier = modifier
            .semantics(mergeDescendants = true) { contentDescription = accessibleText }
            .background(accent.copy(alpha = 0.06f), RoundedCornerShape(LumenRadius.Md))
            .border(1.dp, accent.copy(alpha = 0.18f), RoundedCornerShape(LumenRadius.Md))
            .padding(if (density == LumenStatDensity.Compact) LumenSpacing.Md else LumenSpacing.Lg),
        verticalArrangement = Arrangement.spacedBy(
            if (density == LumenStatDensity.Compact) LumenSpacing.Xs else LumenSpacing.Sm
        )
    ) {
        graphic?.invoke()

        Text(
            text = value,
            color = colors.ink,
            style = (if (density == LumenStatDensity.Compact) {
                MaterialTheme.typography.titleMedium
            } else {
                MaterialTheme.typography.headlineSmall
            }).copy(fontWeight = FontWeight.Bold)
        )
        Text(
            text = label,
            color = colors.inkSoft,
            style = MaterialTheme.typography.labelSmall.copy(fontWeight = FontWeight.SemiBold)
        )

        if (detail != null) {
            Text(text = detail, color = colors.inkMuted, style = MaterialTheme.typography.bodySmall)
        }
    }
}

@Composable
fun LumenSectionHeader(
    title: String,
    modifier: Modifier = Modifier,
    subtitle: String? = null,
    count: String? = null,
    actions: (@Composable () -> Unit)? = null
) {
    val colors = LocalLumenTheme.current.colors

    Row(
        modifier = modifier.fillMaxWidth(),
        horizontalArrangement = Arrangement.spacedBy(LumenSpacing.Md),
        verticalAlignment = Alignment.Top
    ) {
        Column(
            modifier = Modifier.weight(1f),
            verticalArrangement = Arrangement.spacedBy(LumenSpacing.Xs)
        ) {
            Row(
                horizontalArrangement = Arrangement.spacedBy(LumenSpacing.Sm),
                verticalAlignment = Alignment.CenterVertically
            ) {
                Text(
                    text = title,
                    modifier = Modifier.semantics { heading() },
                    color = colors.ink,
                    style = MaterialTheme.typography.titleMedium.copy(fontWeight = FontWeight.SemiBold)
                )
                if (count != null) LumenBadge(count)
            }

            if (subtitle != null) {
                Text(text = subtitle, color = colors.inkMuted, style = MaterialTheme.typography.bodySmall)
            }
        }

        actions?.invoke()
    }
}

@Composable
fun LumenStatusBar(
    message: String,
    modifier: Modifier = Modifier,
    tone: LumenMetricTone = LumenMetricTone.Neutral,
    iconName: LumenIconName? = null,
    trailing: (@Composable () -> Unit)? = null
) {
    val colors = LocalLumenTheme.current.colors

    Row(
        modifier = modifier
            .fillMaxWidth()
            .heightIn(min = 36.dp)
            .background(colors.surfaceMuted)
            .padding(horizontal = LumenSpacing.Lg),
        horizontalArrangement = Arrangement.spacedBy(LumenSpacing.Sm),
        verticalAlignment = Alignment.CenterVertically
    ) {
        LumenIcon(
            name = iconName ?: lumenStatusBarIconName(tone),
            size = LumenIconSize.Sm,
            tint = lumenMetricColor(colors, tone)
        )
        Text(
            text = message,
            modifier = Modifier.weight(1f),
            color = colors.inkSoft,
            style = MaterialTheme.typography.bodySmall
        )
        trailing?.invoke()
    }
}
